### PR TITLE
[Quantization] Support model-defined HF quantized checkpoint loading

### PR DESCRIPTION
**Pull request by Lmywl**
#### Before submitting

- [x] Lint code. `pre-commit run --files <changed files>` passes (black / isort / flake8 / copyright_checker).
- [x] Add test cases into `tests` folder — `tests/ai_edited_test/quantization/test_ai_hf_checkpoint_dequant.py`, 69 cases.

### PR types

New features

### PR changes

APIs

### Description

> [!IMPORTANT]
> **Depends on PaddlePaddle/Paddle#79712** — *[FlexCheckpoint] Add LoadTransform extension point for quantized checkpoints*.
>
> This PR calls `dist.load_state_dict(..., load_transform=...)`; that keyword argument only exists once #79712 lands. CI here will fail until then, which is why this is opened as a **draft**.

#### Motivation

Loading a quantized HuggingFace checkpoint (DeepSeek-V4 FP8, Kimi-K3 MXFP4) currently requires an offline dequantize-and-rewrite pass, which doubles checkpoint storage and adds a manual step before every experiment. This PR lets FlexCheckpoint read the quantized files directly and dequantize during the read.

#### Design

Quantization rules belong to the model definition, not the loader. Checkpoint files only carry physical tensor metadata (names, shapes, dtypes); they do not say which tensors are quantized, how the scales are encoded, or what the block geometry is. So a model exposes a parameterless descriptor:

```python
@classmethod
def _gen_hf_quan_config(cls):
    return {
        "schema_version": 1,
        "component_pairing": {"weight_suffix": ".weight", "scale_suffix": ".scale"},
        "logic_name_suffix": ".weight",
        "groups": [
            {
                "name": ...,
                "targets": [...],          # literal names or "re:" patterns
                "quant_method": ...,       # fp8_block | mxfp4_group
                "value_format": ...,       # e4m3 | e2m1
                "scale_format": ...,       # float32 | ue8m0
                "block_shape": [...],
            }
        ],
    }
```

The descriptor is attached to the built model in the provider factory, so the `Trainer` only needs a `getattr` — a model without the descriptor keeps the old behaviour and is unaffected.

`Trainer` builds a `LoadTransform` from it and hands it to FlexCheckpoint:

```python
if self.args.load_from_hf:
    hf_quan_config = None
    if self.args.hf_quantized_load_mode:
        gen = getattr(self.model, "_gen_hf_quan_config", None)
        if callable(gen):
            hf_quan_config = gen()
    load_transform = build_hf_dequant_load_transform(
        checkpoint_path=resume_from_checkpoint,
        mode=self.args.hf_quantized_load_mode,
        quan_config=hf_quan_config,
    )
    dist.load_state_dict(..., load_transform=load_transform)
```

#### Changes

| File | Change |
| --- | --- |
| `paddleformers/quantization/hf_checkpoint.py` | new — `QuanDescriptor`, `QuanMetadata`, `HFDequantLoadTransform`, `build_hf_dequant_load_transform` |
| `paddleformers/quantization/checkpoint_dequant.py` | new — `FP8BlockCheckpointDequantizer`, `MXFP4GroupCheckpointDequantizer`, plus a `register_checkpoint_dequantizer` hook for new methods |
| `paddleformers/quantization/__init__.py` | export `build_hf_dequant_load_transform` |
| `paddleformers/trainer/training_args.py` | new `hf_quantized_load_mode` argument (default `False`) |
| `paddleformers/trainer/trainer.py` | build the load transform and pass it to `dist.load_state_dict` |
| `paddleformers/transformers/kimi_k3/modeling.py` | `_gen_hf_quan_config` — `mxfp4_group`, `e2m1` values, `ue8m0` scales, `block_shape=[32]` |
| `paddleformers/transformers/deepseek_v4/modeling.py` | `_gen_hf_quan_config` — `fp8_block`, `e4m3` values, `float32` scales, `block_shape=[128, 128]` |
| `tests/ai_edited_test/quantization/test_ai_hf_checkpoint_dequant.py` | new — 69 cases |

Behaviour is fully opt-in: with `hf_quantized_load_mode=False` (the default) no descriptor is consulted and the load path is byte-for-byte the previous one.

#### Validation against a real checkpoint

The DeepSeek-V4 descriptor was checked against `DeepSeek-V4-Flash-Base` (43 layers, 256 routed experts, 1 MTP layer) by parsing all 46 safetensors headers — 69189 tensors total (34167 `F8_E4M3`, 34584 `F32`, 435 `BF16`, 3 `I64`):

- 34167 / 34167 FP8 weights matched — no false negatives, no false positives.
- Every matched weight has its paired `.scale`; no orphan scales.
- Near-miss names are correctly excluded: `attn.compressor.wkv`, `attn.indexer.compressor.{wkv,wgate}` are `BF16` and not matched, while `attn.indexer.wq_b` is `FP8` and is matched.
- `block_shape=[128, 128]` is self-consistent with every weight/scale shape pair.

One note worth recording: the checkpoint's `config.json` declares `scale_fmt: "ue8m0"`, but the stored scales are plain `F32`. The descriptor therefore uses `scale_format: "float32"`, which is what the files actually contain.

#### Tests

`tests/ai_edited_test/quantization/test_ai_hf_checkpoint_dequant.py` — 69 cases, all passing. They cover descriptor parsing and validation, weight/scale pairing, logical-name derivation, both dequantizers (including block/group edge sizes and padding), the registry hook, and error paths. The tests use synthetic descriptors rather than importing any concrete model, so they run without a model checkpoint.

Patch coverage of the changed lines is 92.0% (462/502), above the 75% gate in `.github/codecov.yml`.

### CODE REVIEW

File: paddleformers/quantization/hf_checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,580 @@
...
+        output_dtype: paddle.dtype = paddle.bfloat16,
+    ) -> QuanMetadata:
+        source_keys = _physical_state_dict_metadata(physical_metadata)
+        relations: dict[str, HFQuantizedWeightSpec] = {}
+        groups: dict[str, HFQuantizationGroupSpec] = {}
+        matched_scales: set[str] = set()
+        for weight_key in sorted(source_keys):
+            if not weight_key.endswith(self.weight_suffix):
+                continue
+            matches = [group for group in self._groups if any(pattern.search(weight_key) for pattern in group.targets)]
```
[the remark this answers is not in the document]
Comment (Lmywl) [reply paddleformers/quantization/hf_checkpoint.py:339]: 已修复，commit `7784ed5b`。

字面量 target 改为在**编译期**加锚 —— `_descriptor_patterns` 生成 pattern 时用
`rf"\A{re.escape(target)}\Z"`，字面量因此只能匹配完整名称。

没有把匹配点的 `search()` 换成 `fullmatch()`：那样会一并收紧 `re:` 分支，而 `re:` 形式
本来就允许调用方写不全匹配的正则。在编译期加锚的话，`re:` 分支编译出的 pattern 与改动前
逐字符相同，in-tree 的 DeepSeek-V4 / Kimi-K3 descriptor（各只有一条带 `^...$` 的 `re:`
target）行为完全不变。

顺带补了 `_descriptor_patterns` 的 docstring，说明两种形式的定位：`re:` 是首选形式，
真实 ckpt 的权重按 layer / expert 编号命名，只有正则能描述一整族；裸字符串只作为补充手段，
表示一个完整限定的物理 tensor 名。

负向用例 `test_literal_targets_do_not_match_longer_names` 覆盖三种「前后带额外字符」的情形，
均断言 `build_metadata` 抛错：

- key 多一段前缀：target `layers.0.attn.wq_a.weight` vs key `prefix.layers.0.attn.wq_a.weight`
- 层号共享数字前缀：target `layers.1` vs key `layers.11.attn.wq_a.weight`
- target 短于 weight 后缀：target `layers.0.attn.wq_a` vs key `layers.0.attn.wq_a.weight`

这三条在旧语义（仅 `re.escape` + `search`）下都会命中，改动后都不命中。

另外提一下这条问题的实际危害不止「多纳入一对 weight/scale」：`block_axes` 是从该组第一个
匹配上的 pair 推断一次的，误匹配进来的张量若排在前面，会把整组的 block 几何推歪。

**Review by zjjlivein — DISMISSED** [review]

**Review by From00 — APPROVED** [review]
LGTM

**Comment by Lmywl** [comment]
@Paddle-Bot
